sgit/core.py

```diff
@@ -576,24 +623,30 @@ def _select(self, sequence, selection_query, selection_method):
             elif selection_query == "first":
                 return sequence[0]
             else:
-                spec = SpecifierSet(selection_query)
+                log.debug(f"Selection query")
 
-                filtered_versions = list(
-                    spec.filter(
-                        sequence,
+                try:
+                    spec = SpecifierSet(selection_query)
+
+                    filtered_versions = list(
+                        spec.filter([str(item) for item in sequence]),
                     )
-                )
 
-                log.debug(f"filtered_versions")
-                log.debug(filtered_versions)
+                    log.debug(f"filtered_versions")
+                    log.debug(filtered_versions)
 
-                return filtered_versions[-1]
-        elif selection_method == SelectionMethods.EXACT:
+                    return filtered_versions[-1]
+                except packaging.specifiers.InvalidSpecifier:
+                    log.warning(f"WARNING: Invalid SEMVER select query. Falling back to EXCAT matching of value")
+                    selection_method = SelectionMethods.EXACT
+        
+        if selection_method == SelectionMethods.EXACT:
             for item in sequence:
-                if item == selection_query:
+                if str(item) == selection_query:
                     return item
 
             # Query not found in sequence, return None
             return None
-        else:
+        
+        if selection_method not in SelectionMethods.__members__:
             raise SgitConfigException(f"Unsupported select algorithm selected")
```

Comment:
```suggestion
        elif selection_method not in SelectionMethods.__members__:
            raise SgitConfigException(f"Unknown select algorithm selected")
```

Reply:
@Grokzen I updated this to add an assertion on expected unreachable code. 

https://stackoverflow.com/questions/815310/how-to-designate-unreachable-python-code

Reply:
Assertions is generally forbidden or discouraged extremely to use in main code. The correct way is to do what was done by raisin an exception. Assertions is made for test code and not regular code imo

Reply:
You are correct in a way. This is the reason why I also attached the SO link to give reason for my suggestion. An assertion is an exception. Also, marking clearly unreachable code like this is not uncommon or bad practice. The alternative would be that this case would not be caught at all, which is even worse.

Reply:
Raising `SgitConfigException` as a last resort is a good solution as well. I just think it is good to be defensive here and not let if/if/if be the only way to catch problems with the code. Future changes might change the logic that you are going for now and therefore it is good to be on the safe side even if that happens.